services/system/Transact/src/RTransact.cpp

```diff
@@ -177,129 +179,327 @@ namespace
+   auto claimClientReply(const psibase::Checksum256& id, const ClientFilter& clientFilter)
+       -> std::tuple<std::vector<std::int32_t>, std::vector<std::int32_t>>
+   {
+      {
+         auto                      clients = RTransact::Subjective{}.open<TraceClientTable>();
+         std::vector<std::int32_t> json_clients, bin_clients;
 
-   TransactionTraceRef pruned = PruneTrace{true}(trace);
+         auto socketClaimed = [&](const auto& client)
+         {
+            return std::ranges::contains(json_clients, client.socket) ||
+                   std::ranges::contains(bin_clients, client.socket);
+         };
 
-   auto                          clients = Subjective{}.open<TraceClientTable>();
-   std::optional<TraceClientRow> row;
-   bool                          json;
-   bool                          bin;
-   PSIBASE_SUBJECTIVE_TX
+         PSIBASE_SUBJECTIVE_TX
+         {
+            if (auto row_opt = clients.get(id))
+            {
+               auto& row = *row_opt;
+               if (std::ranges::none_of(row.clients, clientFilter))
+                  abortMessage("No clients matching filter to claim a reply for this tx");
+
+               auto filtered = row.clients | std::views::filter(clientFilter);
+               for (const auto& client : filtered)
+               {
+                  to<HttpServer>().claimReply(client.socket);
+                  auto& clientList = client.json ? json_clients : bin_clients;
+                  clientList.push_back(client.socket);
+               }
+
+               std::erase_if(row.clients, socketClaimed);
+
+               if (row.clients.empty())
+                  clients.remove(row);
+               else
+                  clients.put(row);
+            }
+         }
+
+         return {std::move(json_clients), std::move(bin_clients)};
+      }
+   }
+
+   auto finalizeBlocks(const psibase::BlockHeader& current)
+       -> std::pair<std::vector<psibase::BlockNum>, psibase::BlockTime>
    {
-      json = false;
-      bin  = false;
-      row  = clients.get(id);
-      if (row)
       {
-         clients.remove(*row);
-         for (auto client : row->clients)
+         auto commitNum  = current.commitNum;
+         auto reversible = RTransact::WriteOnly{}.open<ReversibleBlocksTable>();
+         reversible.put({.blockNum = current.blockNum, .time = current.time});
+
+         BlockTime irreversibleTime = {};
+
+         std::vector<psibase::BlockNum> irreversible;
+         for (auto r : reversible.getIndex<0>())
          {
-            if (client.json)
-               json = true;
-            else
-               bin = true;
-            to<HttpServer>().claimReply(client.socket);
+            if (r.blockNum > commitNum)
+               break;
+            irreversibleTime = r.time;
+            if (r.blockNum < commitNum)
+            {
+               reversible.remove(r);
+               irreversible.push_back(r.blockNum);
```

Comment:
This will delay the responses by one block. A block is irreversible if `r.blockNum <= commitNum`. The reason that the last irreversible block is not removed, is so there will always be at least one irreversible block to provide `irreversibleTime`.

Reply:
Ah, I see. Thanks.
Fixed.